lms/djangoapps/grades/tasks.py

```diff
@@ -36,63 +56,57 @@ def recalculate_subsection_grade(
         score_deleted (boolean): indicating whether the grade change is
             a result of the problem's score being deleted.
     """
-    course_key = CourseLocator.from_string(course_id)
+    course_key = CourseLocator.from_string(kwargs['course_id'])
     if not PersistentGradesEnabledFlag.feature_enabled(course_key):
         return
 
     score_deleted = kwargs['score_deleted']
-    scored_block_usage_key = UsageKey.from_string(usage_id).replace(course_key=course_key)
+    scored_block_usage_key = UsageKey.from_string(kwargs['usage_id']).replace(course_key=course_key)
+    task_queued_time = datetime.strptime(kwargs['task_queued_time'], "%Y-%m-%d %H:%M:%S.%f").replace(tzinfo=pytz.UTC)
 
     # Verify the database has been updated with the scores when the task was
     # created. This race condition occurs if the transaction in the task
     # creator's process hasn't committed before the task initiates in the worker
     # process.
     if not _has_database_updated_with_new_score(
-            user_id, scored_block_usage_key, weighted_earned, weighted_possible, score_deleted,
+            kwargs['user_id'], scored_block_usage_key, task_queued_time, score_deleted,
     ):
-        raise _retry_recalculate_subsection_grade(
-            user_id, course_id, usage_id, only_if_higher, weighted_earned, weighted_possible, score_deleted,
-        )
+        raise _retry_recalculate_subsection_grade(**kwargs)
 
     _update_subsection_grades(
         course_key,
         scored_block_usage_key,
-        only_if_higher,
-        course_id,
-        user_id,
-        usage_id,
-        weighted_earned,
-        weighted_possible,
+        kwargs['only_if_higher'],
+        kwargs['course_id'],
+        kwargs['user_id'],
+        kwargs['usage_id'],
+        kwargs['task_queued_time'],
         score_deleted,
     )
 
 
 def _has_database_updated_with_new_score(
-        user_id, scored_block_usage_key, expected_raw_earned, expected_raw_possible, score_deleted,
+        user_id, scored_block_usage_key, task_queued_time, score_deleted,
 ):
     """
     Returns whether the database has been updated with the
     expected new score values for the given problem and user.
     """
-    score = get_score(user_id, scored_block_usage_key)
+    score = get_full_score(user_id, scored_block_usage_key)
 
     if score is None:
         # score should be None only if it was deleted.
         # Otherwise, it hasn't yet been saved.
         return score_deleted
 
-    found_raw_earned, found_raw_possible = score  # pylint: disable=unpacking-non-sequence
-    return (
-        found_raw_earned == expected_raw_earned and
-        found_raw_possible == expected_raw_possible
-    )
+    # In non-async environments, we're still inside an uncommited transaction. This means that score.modified will be
+    # reporting a time slightly earlier than task_queued_time, so we add a "fudge factor" of 1/10 of a second.
```

Comment:
Rather than addressing it this way, I suggest passing the `modified` value through rather than computing the current timestamp.  This would:
1. avoid atomic time issues.
2. work for both async and sync operations.